src/meta/src/barrier/schedule.rs

```diff
@@ -320,63 +323,209 @@ pub struct ScheduledBarriers {
+        for (db_id, db_state) in &mut self.databases {
+            if db_state.barrier_interval.is_none() {
+                let mut interval = tokio::time::interval(duration);
+                interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
+                self.timer_streams
+                    .insert(*db_id, IntervalStream::new(interval));
+            }
+        }
+    }
+
+    /// Update the system checkpoint frequency.
+    pub fn set_sys_checkpoint_frequency(&mut self, frequency: u64) {
+        if self.sys_checkpoint_frequency == frequency {
+            return;
+        }
+        self.sys_checkpoint_frequency = frequency;
+        // Reset the `num_uncheckpointed_barrier` for all databases that use default param.
+        for db_state in self.databases.values_mut() {
+            if db_state.checkpoint_frequency.is_none() {
+                db_state.num_uncheckpointed_barrier = 0;
+                db_state.force_checkpoint = false;
+            }
+        }
+    }
+
+    pub(super) fn update_database_barrier(
+        &mut self,
+        database_id: DatabaseId,
+        barrier_interval_ms: Option<u32>,
+        checkpoint_frequency: Option<u64>,
+    ) {
+        match self.databases.entry(database_id) {
+            Entry::Occupied(mut entry) => {
+                let db_state = entry.get_mut();
+                db_state.barrier_interval =
+                    barrier_interval_ms.map(|ms| Duration::from_millis(ms as u64));
+                db_state.checkpoint_frequency = checkpoint_frequency;
+                // Reset the `num_uncheckpointed_barrier` since the barrier interval or checkpoint frequency is changed.
+                db_state.num_uncheckpointed_barrier = 0;
+                db_state.force_checkpoint = false;
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(DatabaseBarrierState::new(
+                    barrier_interval_ms,
+                    checkpoint_frequency,
+                ));
+            }
         }
+
+        // If the database already has a timer stream, reset it with the new interval.
+        let duration = if let Some(ms) = barrier_interval_ms {
+            Duration::from_millis(ms as u64)
+        } else {
+            self.sys_barrier_interval
+        };
+        self.timer_streams.insert(
+            database_id,
+            IntervalStream::new(tokio::time::interval(duration)),
+        );
     }
 
-    pub(super) fn set_min_interval(&mut self, min_interval: Duration) {
-        let set_new_interval = min_interval != self.min_interval.period();
-        if set_new_interval {
-            let mut min_interval = tokio::time::interval(min_interval);
-            min_interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
-            self.min_interval = min_interval;
+    /// Make the `checkpoint` of the next barrier must be true.
+    pub fn force_checkpoint_in_next_barrier(&mut self) {
```

Comment:
Seems this method is called only within the scope of `DatabaseCheckpointControl`, which means we can specify the `database_id` we want to force checkpoint on, instead of all databases.